# Setting up the notebook

In [1]:
from google.colab import files
import pandas as pd
import io
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split

In [2]:
upload = files.upload()

Saving roundfive.csv to roundfive.csv
Saving roundfour.csv to roundfour.csv
Saving roundone.csv to roundone.csv
Saving roundthree.csv to roundthree.csv
Saving roundtwo.csv to roundtwo.csv


In [3]:
df1 = pd.read_csv(io.BytesIO(upload['roundone.csv']))
df2 = pd.read_csv(io.BytesIO(upload['roundtwo.csv']))
df3 = pd.read_csv(io.BytesIO(upload['roundthree.csv']))
df4 = pd.read_csv(io.BytesIO(upload['roundfour.csv']))
df5 = pd.read_csv(io.BytesIO(upload['roundfive.csv']))
print(df1)

    Unnamed: 0  team_id  match_id            team.name       away_team_name  \
0            1      859   3835319      Austria Women's      Austria Women's   
1            2      865   3835319      England Women's      Austria Women's   
2            3     2457   3835320     Northern Ireland     Northern Ireland   
3            4      852   3835320       Norway Women's     Northern Ireland   
4            5      863   3835321        Spain Women's          WNT Finland   
5            6      920   3835321          WNT Finland          WNT Finland   
6            7      853   3835322      Denmark Women's      Denmark Women's   
7            8      857   3835322      Germany Women's      Denmark Women's   
8            9      864   3835323     Portugal Women's  Switzerland Women's   
9           10      860   3835323  Switzerland Women's  Switzerland Women's   
10          11      851   3835324  Netherlands Women's       Sweden Women's   
11          12      858   3835324       Sweden Women

# Training the models

## Round 1

Organize Round 1 Data


In [4]:
# organize X and y dataframes

X1 = df1.loc[:, "percentfirsttime":"numshotassist"]
print(X1.info())
y1 = df1.loc[:, "outcome"]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16 entries, 0 to 15
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   percentfirsttime  16 non-null     float64
 1   duration_perc     16 non-null     float64
 2   num_crosses       16 non-null     int64  
 3   Total_shots       16 non-null     int64  
 4   Ratio_of_Shots    16 non-null     float64
 5   numshotassist     16 non-null     int64  
dtypes: float64(3), int64(3)
memory usage: 896.0 bytes
None


Tune Round 1 RF

In [ ]:
param_grid = {
    'bootstrap': [True, False],
    'n_estimators': [100,200,300,400,500],
    'max_depth' : [3,5,None],
    'max_features': [n for n in range(1,7)]
}
# Create a based model
rf1 = RandomForestClassifier()
# Instantiate the grid search model
grid_search1 = GridSearchCV(estimator = rf1, param_grid = param_grid, 
                          cv = 3)

# Fit the grid search to the data
grid_search1.fit(X1, y1)
print(grid_search1.best_params_)

{'bootstrap': False, 'max_depth': 3, 'max_features': 4, 'n_estimators': 100}


Fit Round 1 RF

In [5]:
clf1 = RandomForestClassifier(n_estimators = 100, bootstrap = False, max_depth = 3, max_features = 4)
clf1.fit(X1, y1)

train1_pred = clf1.predict(X1)
accuracy1 = metrics.accuracy_score(y1, train1_pred)
print("Round 1 train accuracy: ", accuracy1)

Round 1 train accuracy:  1.0


## Round 2

Organize Round 2 Data

In [6]:
# organize X and y dataframes

X2 = df2.loc[:, "percentfirsttime":"numshotassist"]
print(X2.info())
y2 = df2.loc[:, "outcome"]

# only round 2 outcomes
y2only = df2.loc[16:, "outcome"]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   percentfirsttime  32 non-null     float64
 1   duration_perc     32 non-null     float64
 2   num_crosses       32 non-null     int64  
 3   Total_shots       32 non-null     int64  
 4   Ratio_of_Shots    32 non-null     float64
 5   numshotassist     32 non-null     int64  
dtypes: float64(3), int64(3)
memory usage: 1.6 KB
None


Tune Round 2 RF

In [ ]:
# Create a based model
rf2 = RandomForestClassifier()
# Instantiate the grid search model
grid_search2 = GridSearchCV(estimator = rf2, param_grid = param_grid, 
                          cv = 3)

# Fit the grid search to the data
grid_search2.fit(X2, y2)
print(grid_search2.best_params_)

{'bootstrap': True, 'max_depth': None, 'max_features': 6, 'n_estimators': 100}


Fit Round 2 RF

In [7]:
clf2 = RandomForestClassifier(n_estimators = 100, bootstrap = True, max_depth = 6, max_features = 6)
clf2.fit(X2, y2)

train2_pred = clf2.predict(X2)
accuracy2 = metrics.accuracy_score(y2, train2_pred)
print("Round 2 train accuracy: ", accuracy2)

Round 2 train accuracy:  1.0


## Round 3

Organize Round 3 Data

In [8]:
# organize X and y dataframes

X3 = df3.loc[:, "percentfirsttime":"numshotassist"]
print(X3.info())
y3 = df3.loc[:, "outcome"]

# only round 3 outcomes
y3only = df3.loc[32:, "outcome"]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48 entries, 0 to 47
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   percentfirsttime  48 non-null     float64
 1   duration_perc     48 non-null     float64
 2   num_crosses       48 non-null     int64  
 3   Total_shots       48 non-null     int64  
 4   Ratio_of_Shots    48 non-null     float64
 5   numshotassist     48 non-null     int64  
dtypes: float64(3), int64(3)
memory usage: 2.4 KB
None


Tune Round 3 RF

In [ ]:
# Create a based model
rf3 = RandomForestClassifier()
# Instantiate the grid search model
grid_search3 = GridSearchCV(estimator = rf3, param_grid = param_grid, 
                          cv = 3)

# Fit the grid search to the data
grid_search3.fit(X3, y3)
print(grid_search3.best_params_)

{'bootstrap': True, 'max_depth': None, 'max_features': 5, 'n_estimators': 200}


Fit Round 3 RF

In [9]:
clf3 = RandomForestClassifier(n_estimators = 200, bootstrap = True, max_depth = None, max_features = 5)
clf3.fit(X3, y3)

train3_pred = clf3.predict(X3)
accuracy3 = metrics.accuracy_score(y3, train3_pred)
print("Round 3 train accuracy: ", accuracy3)

Round 3 train accuracy:  1.0


## Round 4

Organize Round 4 Data

In [10]:
# organize X and y dataframes

X4 = df4.loc[:, "percentfirsttime":"numshotassist"]
print(X4.info())
y4 = df4.loc[:, "outcome"]

# only round 4 outcomes
y4only = df4.loc[48:, "outcome"]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   percentfirsttime  56 non-null     float64
 1   duration_perc     56 non-null     float64
 2   num_crosses       56 non-null     int64  
 3   Total_shots       56 non-null     int64  
 4   Ratio_of_Shots    56 non-null     float64
 5   numshotassist     56 non-null     int64  
dtypes: float64(3), int64(3)
memory usage: 2.8 KB
None


Tune Round 4 RF

In [ ]:
# Create a based model
rf4 = RandomForestClassifier()
# Instantiate the grid search model
grid_search4 = GridSearchCV(estimator = rf4, param_grid = param_grid, 
                          cv = 3)

# Fit the grid search to the data
grid_search4.fit(X4, y4)
print(grid_search4.best_params_)

{'bootstrap': True, 'max_depth': 5, 'max_features': 6, 'n_estimators': 400}


Fit Round 4 RF

In [ ]:
clf4 = RandomForestClassifier(n_estimators = 400, bootstrap = True, max_depth = 5, max_features = 6)
clf4.fit(X4, y4)

train4_pred = clf4.predict(X4)
accuracy4 = metrics.accuracy_score(y4, train4_pred)
print("Round 4 train accuracy: ", accuracy4)

Round 4 train accuracy:  1.0


## Round 5

Organize Round 5 Data

In [11]:
# organize X and y dataframes

X5 = df5.loc[:, "percentfirsttime":"numshotassist"]
print(X5.info())
y5 = df5.loc[:, "outcome"]

# only round 5 outcomes
y5only = df5.loc[56:, "outcome"]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   percentfirsttime  60 non-null     float64
 1   duration_perc     60 non-null     float64
 2   num_crosses       60 non-null     int64  
 3   Total_shots       60 non-null     int64  
 4   Ratio_of_Shots    60 non-null     float64
 5   numshotassist     60 non-null     int64  
dtypes: float64(3), int64(3)
memory usage: 2.9 KB
None


Tune Round 5 RF

In [ ]:
# Create a based model
rf5 = RandomForestClassifier()
# Instantiate the grid search model
grid_search5 = GridSearchCV(estimator = rf5, param_grid = param_grid, 
                          cv = 3)

# Fit the grid search to the data
grid_search5.fit(X5, y5)
print(grid_search5.best_params_)

{'bootstrap': True, 'max_depth': 3, 'max_features': 1, 'n_estimators': 400}


Fit Round 5 RF

In [12]:
clf5 = RandomForestClassifier(n_estimators = 400, bootstrap = True, max_depth = 3, max_features = 1)
clf5.fit(X5, y5)

train5_pred = clf5.predict(X5)
accuracy5 = metrics.accuracy_score(y5, train5_pred)
print("Round 5 train accuracy: ", accuracy5)

Round 5 train accuracy:  0.8166666666666667


# Example Predicting with the models

In [ ]:
upload = files.upload()

Organize data frame of previous averages

In [ ]:
"""
Upload data frame containing 16 rows (1 for each team) "averaging" features from previous rounds.
The order of the teams should align with the games from round 2 data

Data frame info should look like below:

RangeIndex: 16 entries, 0 to 15
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   percentfirsttime  16 non-null     float64
 1   duration_perc     16 non-null     float64
 2   num_crosses       16 non-null     float64  
 3   Total_shots       16 non-null     float64  
 4   Ratio_of_Shots    16 non-null     float64
 5   numshotassist     16 non-null     float64 

"""
X1_avgs = pd.read_csv(io.BytesIO(upload['roundoneavgs.csv']))
print(X1_avgs)

Predicting Round 2 using CLF trained on Round 1

In [ ]:
y2_pred = clf1.predict(X1_avgs)
accuracy5 = metrics.accuracy_score(y2only, y2_pred)
print("Round 2 accuracy: ", accuracy2)

# Final Match Prediction

## Based on Means of Previous Games

In [67]:
upload = files.upload()

Saving finalmatch.csv to finalmatch (4).csv


In [68]:
# df containing means of previous game features for England and Germany
dffinal = pd.read_csv(io.BytesIO(upload['finalmatch.csv']))
dffinal

team.name  percentfirsttime  duration_perc  num_crosses  Total_shots  \
0  England Women's          0.386310       0.606648         16.6         19.4   
1  Germany Women's          0.236256       0.538152         13.8         20.2   

   Ratio_of_Shots  numshotassist  
0        3.164967           11.4  
1        7.934762           10.8

In [69]:
X_final = dffinal.iloc[:, 1:]

In [70]:
# row 0: England means, row 1: Germany means
final_pred = clf5.predict(X_final)
print(final_pred)
final_proba = clf5.predict_proba(X_final)
print(final_proba)

eng_avg_win_proba = final_proba[0][2]
ger_avg_win_proba = final_proba[1][2]

['Win' 'Win']
[[0.1097226  0.10502589 0.78525151]
 [0.15252529 0.10832616 0.73914856]]


England average win probability: **0.78525151**

Germany average win probability: **0.73914856**

**Actual Results:** England 2 vs Germany 1

## Variations in Team Perfomance from Previous Games

In [42]:
upload = files.upload()

Saving syndata_eng.csv to syndata_eng (1).csv
Saving syndata_ger.csv to syndata_ger (1).csv


In [43]:
dfeng = pd.read_csv(io.BytesIO(upload['syndata_eng.csv']))
dfger = pd.read_csv(io.BytesIO(upload['syndata_ger.csv']))

In [44]:
dfeng.info()
dfger.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17 entries, 0 to 16
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              17 non-null     object 
 1   percentfirsttime  17 non-null     float64
 2   duration_perc     17 non-null     float64
 3   num_crosses       17 non-null     float64
 4   Total_shots       17 non-null     float64
 5   Ratio_of_Shots    17 non-null     float64
 6   numshotassist     17 non-null     float64
dtypes: float64(6), object(1)
memory usage: 1.1+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17 entries, 0 to 16
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name_GER          17 non-null     object 
 1   percentfirsttime  17 non-null     float64
 2   duration_perc     17 non-null     float64
 3   num_crosses       17 non-null     float64
 4   Total_shots       17 non-null     

Calculate England Variation Win Probabilities

In [55]:
X_eng = dfeng.iloc[:, 1:]
eng_variations = dfeng.iloc[:, :1]

eng_pred = clf5.predict(X_eng)
eng_variations = eng_variations.assign(result = eng_pred)
eng_proba = clf5.predict_proba(X_eng)
eng_win_proba = [row[2] for row in eng_proba]
eng_variations = eng_variations.assign(win_proba = eng_win_proba)

eng_variations

Name result  win_proba
0            Means    Win   0.785252
1         Inc_by5p    Win   0.816144
2        Inc_by10p    Win   0.804506
3         Dec_by5p    Win   0.768928
4        Dec_by10p    Win   0.764390
5        Inc_by1SD    Win   0.881218
6        Inc_by2SD    Win   0.863987
7       IncF1_by5p    Win   0.776049
8       IncF2_by5p    Win   0.790991
9       IncF3_by5p    Win   0.784064
10      IncF4_by5p    Win   0.783750
11      IncF5_by5p    Win   0.785252
12      IncF6_by5p    Win   0.820172
13    IncF256_by5p    Win   0.827411
14   IncF2456_by5p    Win   0.826534
15   IncF256_by10p    Win   0.824437
16  IncF2456_by10p    Win   0.823560

Calculate Germany Variation Win Probabilities

In [57]:
X_ger = dfger.iloc[:, 1:]
ger_variations = dfger.iloc[:, :1]

ger_pred = clf5.predict(X_ger)
ger_variations = ger_variations.assign(result = ger_pred)
ger_proba = clf5.predict_proba(X_ger)
ger_win_proba = [row[2] for row in ger_proba]
ger_variations = ger_variations.assign(win_proba = ger_win_proba)

ger_variations

Name_GER result  win_proba
0            Means    Win   0.739149
1         Inc_by5p    Win   0.749924
2        Inc_by10p    Win   0.793126
3         Dec_by5p    Win   0.681890
4        Dec_by10p    Win   0.542955
5        Inc_by1SD    Win   0.880557
6        Inc_by2SD    Win   0.856670
7       IncF1_by5p    Win   0.747613
8       IncF2_by5p    Win   0.737456
9       IncF3_by5p    Win   0.739149
10      IncF4_by5p    Win   0.739149
11      IncF5_by5p    Win   0.739149
12      IncF6_by5p    Win   0.743808
13    IncF256_by5p    Win   0.739729
14   IncF2456_by5p    Win   0.739729
15   IncF256_by10p    Win   0.762482
16  IncF2456_by10p    Win   0.774068

Variations for Germany Win with Average England performance

In [60]:
# filtering for higher probability than England average win probability
ger_win_df = ger_variations[ger_variations['win_proba'] > eng_avg_win_proba] 
ger_win_df

Name_GER result  win_proba
2  Inc_by10p    Win   0.793126
5  Inc_by1SD    Win   0.880557
6  Inc_by2SD    Win   0.856670

Variations for England Loss with Average Germany Performance

In [61]:
# filtering for lower probability than Germany average win probability
eng_loss_df = eng_variations[eng_variations['win_proba'] < ger_avg_win_proba] 
eng_loss_df 

Empty DataFrame
Columns: [Name, result, win_proba]
Index: []

## Compare with Actual Game Statistics

In [74]:
upload = files.upload()

Saving actualfinalmatch.csv to actualfinalmatch (3).csv


In [81]:
# df containing final game features for England and Germany
dffinal_actual = pd.read_csv(io.BytesIO(upload['actualfinalmatch.csv']))
dffinal_actual

team.name  percentfirsttime  duration_perc  num_crosses  Total_shots  \
0  England Women's          0.380228       0.518388           14           13   
1  Germany Women's          0.272183       0.481612           21           15   

   Ratio_of_Shots  numshotassist outcome  
0        0.866667              8     Win  
1        1.153846             10    Loss

In [80]:
X_final_actual = dffinal_actual.iloc[:, 1:7]

# row 0: England game stats, row 1: Germany game stats
final_actual_pred = clf5.predict(X_final_actual)
print(final_actual_pred)
final_actual_proba = clf5.predict_proba(X_final_actual)
print(final_actual_proba)

eng_actual_win_proba = final_actual_proba[0][2]
ger_actual_win_proba = final_actual_proba[1][2]

['Win' 'Win']
[[0.2858701  0.22437477 0.48975513]
 [0.25154188 0.17499551 0.57346261]]


England game win probability: **0.48975513**

Germany game win probability: **0.57346261**


Comparing Average Feature with Actual Feature Stats

In [86]:
# row 0: England, row 1: Germany
X_feature_ratios = X_final_actual / X_final
X_feature_ratios # represents the ratio of actual game feature to average previous features

percentfirsttime  duration_perc  num_crosses  Total_shots  Ratio_of_Shots  \
0          0.984258       0.854512     0.843373     0.670103        0.273831   
1          1.152070       0.894938     1.521739     0.742574        0.145417   

   numshotassist  
0       0.701754  
1       0.925926